### PR TITLE
Added sequence settings based on picklist

### PR DESCRIPTION
**Pull request by adudekyeti**
Added sequence settings based on picklist
sequence

Then:
image

### CODE REVIEW

File: layouts/basic/modules/Settings/Vtiger/CustomRecordNumberingAdvanced.tpl [hunk 1]
```diff
@@ -0,0 +1,57 @@
+{*<!-- {[The file is published on the basis of YetiForce Public License 3.0 that can be found in the following directory: licenses/LicenseEN.txt or yetiforce.com]} -->*}
+<div class="tpl-Settings-Vtiger-CustomRecordNumberingAdvanced">
```
Comment (mariuszkrzaczkowski) [inline layouts/basic/modules/Settings/Vtiger/CustomRecordNumberingAdvanced.tpl]:
```suggestion
<div class="tpl-Settings-Base-CustomRecordNumberingAdvanced">
```

File: modules/Settings/Vtiger/actions/CustomRecordNumberingAjax.php [hunk 2]
```diff
@@ -90,12 +93,42 @@ public function saveModuleCustomNumberingData(\App\Request $request)
...
 
+	/**
+	 * Function save module advanced numbering data.
+	 *
+	 * @param App\Request $request
+	 */
+	public function saveModuleCustomNumberingAdvanceData(App\Request $request)
+	{
+		$updated = false;
+		$qualifiedModuleName = $request->getByType('sourceModule', 2);
```
Comment (mariuszkrzaczkowski) [inline modules/Settings/Vtiger/actions/CustomRecordNumberingAjax.php]:
```suggestion
		$sourceModule = $request->getByType('sourceModule', 2);
```

File: modules/Settings/Vtiger/actions/CustomRecordNumberingAjax.php [hunk 3]
```diff
@@ -90,12 +93,42 @@ public function saveModuleCustomNumberingData(\App\Request $request)
...
+		$moduleId = \App\Module::getModuleId($qualifiedModuleName);
+		$picklistValues = $request->getArray('sequenceNumber');
```
Comment (mariuszkrzaczkowski) [inline modules/Settings/Vtiger/actions/CustomRecordNumberingAjax.php]:
```suggestion
		$picklistValues = $request->getMultiDimensionArray('sequenceNumber'[......]);
```

File: modules/Settings/Vtiger/actions/CustomRecordNumberingAjax.php [hunk 4]
```diff
@@ -90,12 +93,42 @@ public function saveModuleCustomNumberingData(\App\Request $request)
 		$response->emit();
 	}
 
+	/**
+	 * Function save module advanced numbering data.
+	 *
+	 * @param App\Request $request
+	 */
+	public function saveModuleCustomNumberingAdvanceData(App\Request $request)
```
Comment (rskrzypczak) [inline modules/Settings/Vtiger/actions/CustomRecordNumberingAjax.php]: ?

File: modules/Settings/Vtiger/views/CustomRecordNumbering.php [hunk 5]
```diff
@@ -17,22 +17,27 @@ class Settings_Vtiger_CustomRecordNumbering_View extends Settings_Vtiger_Index_V
...
 		if ($sourceModule) {
 			$defaultModuleModel = $supportedModules[\App\Module::getModuleId($sourceModule)];
 		} else {
 			$defaultModuleModel = reset($supportedModules);
 		}
-
+		$picklistFields = [];
+		$picklistsModels = $defaultModuleModel->getFieldsByType(['picklist']);
+		foreach ($picklistsModels as $fieldModel) {
+			if (!empty(array_filter(array_column(\App\Fields\Picklist::getValues($fieldModel->getFieldName()), 'prefix')))) {
```
Comment (mariuszkrzaczkowski) [inline modules/Settings/Vtiger/views/CustomRecordNumbering.php]: add new function \App\Fields\Picklist::prefixExist

File: public_html/layouts/basic/modules/Settings/Vtiger/resources/CustomRecordNumbering.js [hunk 6]
```diff
@@ -216,9 +216,57 @@ $.Class('Settings_CustomRecordNumbering_Js', {}, {
...
+			let regex = new RegExp("{{picklist:([a-z0-9_]+)}}", 'g');
+			if ((prefix).match(regex)){
+				picklistName = prefix;
+			} else if((postfix).match(regex)){
+				picklistName = postfix;
+			}
+			AppConnector.request({
+				module: app.getModuleName(),
+				parent: app.getParentModuleName(),
+				view: "CustomRecordNumberingAdvanced",
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/basic/modules/Settings/Vtiger/resources/CustomRecordNumbering.js]:
```suggestion
				view: 'CustomRecordNumberingAdvanced',
```

File: public_html/layouts/basic/modules/Settings/Vtiger/resources/CustomRecordNumberingAdvanced.js [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+				let progressIndicatorElement = $.progressIndicator({
+					position: 'html',
+					blockInfo: {
+						enabled: true
+					}
+				});
+				AppConnector.request({
+					module: app.getModuleName(),
+					parent: app.getParentModuleName(),
+					action: "CustomRecordNumberingAjax",
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/basic/modules/Settings/Vtiger/resources/CustomRecordNumberingAdvanced.js]:
```suggestion
					action: ;CustomRecordNumberingAjax;,
```

File: public_html/layouts/basic/modules/Settings/Vtiger/resources/CustomRecordNumberingAdvanced.js [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+					mode: "saveModuleCustomNumberingAdvanceData",
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/basic/modules/Settings/Vtiger/resources/CustomRecordNumberingAdvanced.js]:
```suggestion
					mode: ;saveModuleCustomNumberingAdvanceData;,
```

File: public_html/layouts/basic/modules/Settings/Vtiger/resources/CustomRecordNumberingAdvanced.js [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+					sourceModule: sourceModule,
+					sequenceNumber: modalForm.serializeFormData()
+				}).done(function (data) {
+					progressIndicatorElement.progressIndicator({mode: 'hide'});
+					if (data.success === true) {
+						Settings_Vtiger_Index_Js.showMessage({
+							text: app.vtranslate('JS_RECORD_NUMBERING_SAVED_SUCCESSFULLY_FOR') + " " + parentForm.find('option[value="' + sourceModule + '"]').text()
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/basic/modules/Settings/Vtiger/resources/CustomRecordNumberingAdvanced.js]:
```suggestion
							text: app.vtranslate('JS_RECORD_NUMBERING_SAVED_SUCCESSFULLY_FOR') + ' ' + parentForm.find('option[value="' + sourceModule + '"]').text()
```

File: modules/Settings/Vtiger/views/CustomRecordNumberingAdvanced.php [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+/**
+ * Modal for advanced record numbering.
+ *
+ * @package   View
+ *
+ * @copyright YetiForce Sp. z o.o.
+ * @license   YetiForce Public License 3.0 (licenses/LicenseEN.txt or yetiforce.com)
+ * @author    Arkadiusz Dudek <a.dudek@yetiforce.com>
+ */
+class Settings_Vtiger_CustomRecordNumberingAdvanced_View extends Settings_Vtiger_BasicModal_View
```
Comment (rskrzypczak) [inline modules/Settings/Vtiger/views/CustomRecordNumberingAdvanced.php]: ?

File: layouts/basic/modules/Settings/Vtiger/CustomRecordNumberingAdvanced.tpl [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+									data-old-sequence-number="{$PICKLIST_VALUE['cur_id']}"
+									name="{$PICKLIST_VALUE['picklistValue']}"
+									data-validation-engine="validate[required,funcCall[Vtiger_WholeNumber_Validator_Js.invokeValidation]]"/>
+								</td>
+							</tr>
+							{/foreach}
+						</tbody>
+					</table>
+				</div>
+				{include file=\App\Layout::getTemplatePath('Modals/Footer.tpl', $QUALIFIED_MODULE) BTN_SUCCESS='LBL_SAVE' BTN_DANGER='LBL_CANCEL' MODULE=$QUALIFIED_MODULE}
```
Comment (rskrzypczak) [inline layouts/basic/modules/Settings/Vtiger/CustomRecordNumberingAdvanced.tpl]: ?

File: layouts/basic/modules/Settings/Vtiger/CustomRecordNumbering.tpl [hunk 12]
```diff
@@ -225,6 +225,14 @@
 								</select>
 							</td>
 						</tr>
+						<tr>
+							<td colspan="2">
+								<button class="btn btn-success float-right js-adavanced-sequence" type="button"
```
Comment (rskrzypczak) [inline layouts/basic/modules/Settings/Vtiger/CustomRecordNumbering.tpl]: ?

File: modules/Settings/Vtiger/actions/CustomRecordNumberingAjax.php (hunk 3 above)
Comment (adudekyeti) [reply modules/Settings/Vtiger/actions/CustomRecordNumberingAjax.php]: ?

File: app/Fields/RecordNumber.php [hunk 13]
```diff
@@ -139,23 +139,38 @@ public function parseNumber(int $seq): string
...
+			$this->updateNumberSequence($reqNo, $this->getPicklistValue($piclistName));
 		}
 		\App\Db::getInstance()->createCommand()->update('vtiger_modentity_num', $data, ['tabid' => $this->get('tabid')])->execute();
 	}
 
+	/**
+	 * Update number sequence.
+	 *
+	 * @param int    $reqNo
+	 * @param string $value
```
Comment (rskrzypczak) [inline app/Fields/RecordNumber.php]: ?

File: modules/Settings/Vtiger/views/CustomRecordNumberingAdvanced.php [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+ */
+class Settings_Vtiger_CustomRecordNumberingAdvanced_View extends \App\Controller\ModalSettings
+{
+	/**
+	 * {@inheritdoc}
+	 */
+	public function preProcessAjax(App\Request $request)
+	{
+		$this->qualifiedModuleName = $request->getModule(false);
+		$this->pageTitle = \App\Language::translate('LBL_ADVANCED_RECORD_NUMBERING', $this->qualifiedModuleName);
```
Comment (rskrzypczak) [inline modules/Settings/Vtiger/views/CustomRecordNumberingAdvanced.php]:
```suggestion
		$this->pageTitle = \App\Language::translate('LBL_ADVANCED_RECORD_NUMBERING', $request->getModule(false));
```

File: modules/Settings/Vtiger/views/CustomRecordNumberingAdvanced.php [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+		$moduleModel = \Vtiger_Module_Model::getInstance($sourceModule);
+		$picklistSequences = (new \App\Db\Query())->select(['value', 'cur_id'])->from('u_#__modentity_sequences')->where(['tabid' => $moduleModel->getId()])->createCommand()->queryAllByGroup(0);
+		$picklistName = $picklist = [];
+		preg_match('/{{picklist:([a-z0-9_]+)}}/i', $request->getByType('picklist', 'Text'), $picklistName);
+		if ($sourceModule) {
+			$defaultModuleModel = $supportedModules[\App\Module::getModuleId($sourceModule)];
+		} else {
+			$defaultModuleModel = reset($supportedModules);
+		}
+		if ($moduleModel->getFieldByName($picklistName[1]) && !empty($picklistName[1])) {
```
Comment (rskrzypczak) [inline modules/Settings/Vtiger/views/CustomRecordNumberingAdvanced.php]:
```suggestion
		if (!empty($picklistName[1]) && $moduleModel->getFieldByName($picklistName[1])) {
```

**Pull request merged**